### PR TITLE
Pachaturvedi/elastic new version 2022 09 01 preview

### PR DESCRIPTION
**Pull request by pachaturvedi**
## ARM API Information (Control Plane)

<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow.
</i>
Azure 1st Party Service can try out the [Shift Left](https://aka.ms/ShiftLeft) experience to initiate API design review from ADO code repo.  If you are interested, may request engineering support by filling in with the form https://aka.ms/ShiftLeftSupportForm.

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [x] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [x] Other, please clarify : Sync from azure-rest-api-specs-pr
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month. December 2022
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month. December 2022
  4. By default, Azure SDKs of all languages (.NET/Python/Java/JavaScript for both management-plane SDK and data-plane SDK, Go for management-plane SDK only ) MUST be refreshed with/after swagger of new version is published. If you prefer NOT to refresh any specific SDK language upon swagger updates in the current PR, please leave details with justification here.

### Contribution checklist (MS Employees Only):
- [x] I commit to follow the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist

> **Applicability**: :warning:
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs
> - Adding new properties
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [x] Check this box if any of the following apply to the PR so that the label "ARMReview" and "WaitForARMFeedback" will be added by bot to kick off ARM API Review.  Missing to check this box in the following scenario may result in delays to the ARM manifest review and deployment.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    - [x] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits. You can use OpenAPIHub to initialize the PR for adding a new version. For more details refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [x] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them.

### Breaking Change Review Checklist
If you have any breaking changes as defined in the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy/), request approval from the Breaking Change Review Board.

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Additional details on the process and office hours are on the [Breaking Change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

NOTE: To update API(s) in public preview for over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

**Comment by pachaturvedi** [comment]
@dapzhang Regarding the breaking changes, this is a swagger fix from our side because we did not have a 202 response for all x-ms-long-running-operation: true (as mentioned in [link](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/async-api-reference.md#202-accepted-and-location-headers) )

**Comment by pachaturvedi** [comment]
@ArcturusZhang could you please help review this PR?

File: specification/elastic/resource-manager/Microsoft.Elastic/preview/2022-09-01-preview/elastic.json
Comment (rkmanda) [inline specification/elastic/resource-manager/Microsoft.Elastic/preview/2022-09-01-preview/elastic.json:1087]: >Success [](http://example.com/codeflow?start=28&length=7)

pl describe the schema of the final response of the 202 using a 200 status code. Also describe the final state using x-ms-long-running-operation-options.

Comment (rkmanda) [reply specification/elastic/resource-manager/Microsoft.Elastic/preview/2022-09-01-preview/elastic.json:1087]: Do this everywhere

File: specification/elastic/resource-manager/Microsoft.Elastic/preview/2022-09-01-preview/elastic.json
Comment (rkmanda) [inline specification/elastic/resource-manager/Microsoft.Elastic/preview/2022-09-01-preview/elastic.json:2269]: This is not a proper way of represeting the API version parameter. Each swagger is already linked ot a single API version. Just use the common type definition for API version instead

**Comment by rkmanda** [comment]
Once you have addressed all feedback provided OR provided a rationale on why the feedback was not addressed, please remove the "ARMChangesRequested" label from the PR.

File: specification/elastic/resource-manager/Microsoft.Elastic/preview/2022-09-01-preview/elastic.json
Comment (pachaturvedi) [reply specification/elastic/resource-manager/Microsoft.Elastic/preview/2022-09-01-preview/elastic.json:1087]: @rkmanda do you mean we should add 200 status code as well along with 201 & 202??

Also, regarding describing the final state using x-ms-long-running-operation-options, this functionality of creating a IP traffic filter is not new, this was introduced in API version 2022-07-01-preview. Could you please let us know if this is really required because PR for previous version when this was introduced was accepted without this.

File: specification/elastic/resource-manager/Microsoft.Elastic/preview/2022-09-01-preview/elastic.json
Comment (pachaturvedi) [reply specification/elastic/resource-manager/Microsoft.Elastic/preview/2022-09-01-preview/elastic.json:2269]: @rkmanda this is how we've been always adding a new version. Could you please point us to the documentation or any reference which describes how this is supposed to be added in the right way?

File: specification/elastic/resource-manager/Microsoft.Elastic/preview/2022-09-01-preview/elastic.json
Comment (pachaturvedi) [reply specification/elastic/resource-manager/Microsoft.Elastic/preview/2022-09-01-preview/elastic.json:1087]: @rkmanda could you please help us here?

File: specification/elastic/resource-manager/Microsoft.Elastic/preview/2022-09-01-preview/elastic.json
Comment (zizw123) [reply specification/elastic/resource-manager/Microsoft.Elastic/preview/2022-09-01-preview/elastic.json:2269]: See common type: [https://github.com/Azure/azure-rest-api-specs-pr/blob/d89e212da2cc34206fe711f44dfcb6f8fdece2a1/specification/common-types/resource-management/v3/types.json#L630](https://github.com/Azure/azure-rest-api-specs-pr/blob/d89e212da2cc34206fe711f44dfcb6f8fdece2a1/specification/common-types/resource-management/v3/types.json#L630)

File: specification/elastic/resource-manager/Microsoft.Elastic/preview/2022-09-01-preview/elastic.json
Comment (zizw123) [reply specification/elastic/resource-manager/Microsoft.Elastic/preview/2022-09-01-preview/elastic.json:1087]: For your final response, what will be returned? Is there no response? And that's why it's returning 201?

Comment (pachaturvedi) [reply specification/elastic/resource-manager/Microsoft.Elastic/preview/2022-09-01-preview/elastic.json:1087]: @zizw123 We return 201 and we don't have any response body. As per the CLI docs 202 response is necessary for all x-ms-long-running-operation: true (as mentioned in [link](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/async-api-reference.md#202-accepted-and-location-headers) ) which was missed in the previous version. Hence we added it in both the versions 2022-07-01-preview & 2022-09-01-preview.

Error message received in CLI
MicrosoftTeams-image (2)

Note: this error is not received when the call is made via Azure Portal or Postman.
Please let me know if you need more info on this, we can have a quick call.

File: specification/elastic/resource-manager/Microsoft.Elastic/preview/2022-09-01-preview/elastic.json
Comment (pachaturvedi) [reply specification/elastic/resource-manager/Microsoft.Elastic/preview/2022-09-01-preview/elastic.json]: Addressed

File: specification/elastic/resource-manager/Microsoft.Elastic/preview/2022-09-01-preview/elastic.json
Comment (zizw123) [reply specification/elastic/resource-manager/Microsoft.Elastic/preview/2022-09-01-preview/elastic.json:1087]: @pachaturvedi As rkmanda mentioned, you should use 200 and 202, see https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/async-api-reference.md#call-action-post-asynchronously

Comment (pachaturvedi) [reply specification/elastic/resource-manager/Microsoft.Elastic/preview/2022-09-01-preview/elastic.json:1087]: @zizw123 update the status code in new and previous version. We'll make the respective changes in our RP code as well.

**Comment by pachaturvedi** [comment]
@zizw123 Addressed all the review comments. Could you please look into this?

**Comment by zizw123** [comment]
> NewApiVersionRequired reason: A service’s API is a contract with customers and is represented by using the api-version query parameter. Changes such as adding an optional property to a request/response or introducing a new operation is a change to the service’s contract and therefore requires a new api-version value. This is critically important for documentation, client libraries, and customer support. EXAMPLE: if a customer calls a service in the public cloud using api-version=2020-07-27, the new property or operation may exist but if they call the service in a government cloud, air-gapped cloud, or Azure Stack Hub cloud using the same api-version, the property or operation may not exist. Because there is no clear relationship between the service api-version and the new property/operation, customers can’t trust the documentation and Azure customer have difficulty helping customers diagnose issues. In addition, each client library version documents the service version it supports. When an optional property or new operation is added to a service and its Swagger, new client libraries must be produced to expose this functionality to customers. Without updating the api-version, it is unclear to customers which version of a client library supports these new features.

Please see this, the new response should be applied to the new api-version, adding it to previous version might break existing users.

**Comment by pachaturvedi** [comment]
> 

@zizw123 this feature is not released to customers yet. It is still under the feature flag. This change was already approved in the previous versions (ref : https://github.com/Azure/azure-rest-api-specs/pull/20791) But since this is not the right way, I've updated it in both the versions and we'll fix this in our RP code base as well so that there is no breaking change across versions. If you feel I can setup a quick call early next week and we can discuss this more on this if required. Kindly let me know.

TIA

**Comment by zizw123** [comment]
> > 
> 
> @zizw123 this feature is not released to customers yet. It is still under the feature flag. This change was already approved in the previous versions (ref : #20791) But since this is not the right way, I've updated it in both the versions and we'll fix this in our RP code base as well so that there is no breaking change across versions. If you feel I can setup a quick call early next week and we can discuss this more on this if required. Kindly let me know.
> 
> TIA

Sounds reasonable to me, but since the PR is tagged with breaking changes, this PR needs to be reviewed by them as well. The rests look good to me. After that is done, I can sign off this PR. Thanks.

**Comment by pachaturvedi** [comment]
**Update**: Following up with Breaking change team for approval - https://msazure.visualstudio.com/One/_workitems/edit/16646257

**Comment by pachaturvedi** [comment]
@zizw123 We've got the Breaking change approval. Could you please approve the PR?

**Comment by pachaturvedi** [comment]
@honghr  Could you please help me with the ~[Staging] Service API Readiness Test . This failure was something which I was not seeing previously.

**Comment by TimLovellSmith** [comment]
@pilor @rkmanda Do the 202 Async patterns here look good to you? I'm also a little confused why a POST action that creates a resource can't return 201 Created plus a Location header pointing at the created resource, as final status. I thought that was a pretty common pattern.

**Comment by pilor** [comment]
If an async POST call is not strictly following https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/async-api-reference.md#call-action-post-asynchronously then its incorrect.  POST calls should not be used to create resources.  While that is used in REST sometimes it is not how Azure APIs work

**Comment by TimLovellSmith** [comment]
@pilor I agree POST calls shouldn't be used to create resources, but seems like in this case it already is (maybe with justification, I'm not sure), so I think the main question is what to do about it in this case - big breaking changes, little breaking changes, or nothing.

**Comment by pilor** [comment]
> @pilor I agree POST calls shouldn't be used to create resources, but seems like in this case it already is (maybe with justification, I'm not sure), so I think the main question is what to do about it in this case - big breaking changes, little breaking changes, or nothing.

thats a question for breaking change folks imo :-)  Sounds like SDKs don't work with the existing design but changing the design is a breaking change

**Comment by utkarshjain1508** [comment]
@rkmanda @TimLovellSmith @honghr @pilor 
Hey folks, addressing the POST design implementation here for traffic filters. The traffic filters are a resource which are not maintained by our service but by an external partner team. Since we do not want to maintain ARM lifecycle for this resource, we have modelled the CRUD operations for this as POST calls. This scenario was addressed when we merged this in the private repo. In case you need more clarification we can discuss this over a call. 
Thanks!
//@pachaturvedi

**Comment by TimLovellSmith** [comment]
@utkarshjain1508 A call sounds like a good idea; just because it is leveraging an external system doesn't necessarily mean you shouldn't use REST verbs. Was there some other reason?

**Comment by pachaturvedi** [comment]
> POST design implementation

Hi @TimLovellSmith, @rkmanda , @pilor  I've setup a call on 01/20/23 9:00 am PST to discuss more on this.  Let me know if that time doesn't works for you.

**Comment by raosuhas** [comment]
Discussed this offline with the team and explained to them the issue with modelling these functions as POST. Main ones being that once these are modelled as POST you lose most benefits that ARM provides like : 
- Deploying the resource with a template 
- Applying the resource as a policy affect or trigger
- Any resource level RBAC
- Potentially storing these in ARG

Once your team has finished the internal discussions on the correct modelling for this please restart the review by removing "ARMChangesRequested" tag. If you decide to keep the current model please have any senior member from your team respond to this comment signing off on this design and understanding that once this API is released it is subject to the azure breaking changes policy for deprecation : http://aka.ms/AzBreakingChangesPolicy

**Comment by karansinghkushwah** [comment]
> Discussed this offline with the team and explained to them the issue with modelling these functions as POST. Main ones being that once these are modelled as POST you lose most benefits that ARM provides like :
> 
> * Deploying the resource with a template
> * Applying the resource as a policy affect or trigger
> * Any resource level RBAC
> * Potentially storing these in ARG
> 
> Once your team has finished the internal discussions on the correct modelling for this please restart the review by removing "ARMChangesRequested" tag. If you decide to keep the current model please have any senior member from your team respond to this comment signing off on this design and understanding that once this API is released it is subject to the azure breaking changes policy for deprecation : http://aka.ms/AzBreakingChangesPolicy

Thanks @raosuhas, @TimLovellSmith for all the feedback here and helping us explore various options. We had multiple rounds of discussion internally within team and with ISV partner. We came to the conclusion that we want to move ahead with the current design approach (Using POST) and we fully understand the limitations. 

Could you please approve the PR if there are no other feedback apart from this.

File: specification/elastic/resource-manager/Microsoft.Elastic/preview/2022-07-01-preview/elastic.json
Comment (pachaturvedi) [reply specification/elastic/resource-manager/Microsoft.Elastic/preview/2022-07-01-preview/elastic.json]: Addressed.

File: specification/elastic/resource-manager/Microsoft.Elastic/preview/2022-09-01-preview/elastic.json
Comment (sjanamma) [inline specification/elastic/resource-manager/Microsoft.Elastic/preview/2022-09-01-preview/elastic.json:1096]: You also need to add x-ms-long-running-operation-options.

[https://github.com/Azure/azure-openapi-validator/blob/main/docs/long-running-operations-options-validator.md](https://github.com/Azure/azure-openapi-validator/blob/main/docs/long-running-operations-options-validator.md)

**Comment by sjanamma** [comment]
Please add x-ms-long-running-operation-options for long running operations

https://github.com/Azure/azure-openapi-validator/blob/main/docs/long-running-operations-options-validator.md
More details here  - https://github.com/Azure/autorest/blob/main/docs/extensions/readme.md#x-ms-long-running-operation

File: specification/elastic/resource-manager/Microsoft.Elastic/preview/2022-09-01-preview/elastic.json
Comment (pachaturvedi) [reply specification/elastic/resource-manager/Microsoft.Elastic/preview/2022-09-01-preview/elastic.json:1096]: Addressed.

**Review by honghr — APPROVED** [review]

**Pull request merged**
